cpp/src/strings/capitalize.cu

```diff
@@ -0,0 +1,309 @@
+    {
+      uint32_t code_point = detail::utf8_to_codepoint(chr);
+      detail::character_flags_table_type flag = code_point <= 0x00FFFF ? d_flags_[code_point] : 0;
+      return char_info{code_point,flag};
+    }
+
+    __device__
+    char_utf8 convert_char(char_info const& info) const
+    {
+      return detail::codepoint_to_utf8(d_case_table_[info.first]);
+    }
+
+    __device__ char* get_output_ptr(size_type idx)
+    {
+      return d_chars_ && d_offsets_ ? d_chars_ + d_offsets_[idx] : nullptr;
+    }
+  private:
+    column_device_view const d_column_;
+    character_flags_table_type const* d_flags_;
+    character_cases_table_type const* d_case_table_;
+    int32_t const* d_offsets_;
+    char* d_chars_;
+  };
+
+
+  //functor for probing string capitalization
+  //requirements:
+  //(private inheritance to prevent polymorphic use,
+  // a requirement that came up in code review)
+  //
+  struct probe_capitalize: private probe_execute_base
+  {
+    explicit probe_capitalize(column_device_view const d_column):
+      probe_execute_base(d_column)
+    {  
+    }
+    
+     __device__
+     int32_t operator()(size_type idx) const {
+       if( get_column().is_null(idx) )
+         return 0; // null string
+      
+       string_view d_str = get_column().template element<string_view>(idx);
+       int32_t bytes = 0;
+      
+       for( auto itr = d_str.begin(); itr != d_str.end(); ++itr ) {
+         auto the_chr = *itr;
+
+         auto pair_char_info = get_char_info(the_chr);
+         detail::character_flags_table_type flag = pair_char_info.second;
+         
+         if( (itr == d_str.begin()) ? IS_LOWER(flag) : IS_UPPER(flag) )
+           the_chr = convert_char(pair_char_info);
+         bytes += detail::bytes_in_char_utf8(the_chr);
+       }
+       return bytes;
+    }
+  };
+
+  //functor for executing string capitalization:
+  //(private inheritance to prevent polymorphic use,
+  // a requirement that came up in code review)
+  //
+  struct execute_capitalize: private probe_execute_base
+  {
+    execute_capitalize(column_device_view const d_column,
+                       int32_t const* d_offsets,
+                       char* d_chars):
+      probe_execute_base(d_column, d_offsets, d_chars)
+    {
+    }
+    
+    __device__
+    int32_t operator()(size_type idx) {
+      if( get_column().is_null(idx) )
+        return 0; // null string
+      
+      string_view d_str = get_column().template element<string_view>(idx);
+      char* d_buffer = get_output_ptr(idx);
+      
+      for( auto itr = d_str.begin(); itr != d_str.end(); ++itr ) {
+        auto the_chr = *itr;
+
+        auto pair_char_info = get_char_info(the_chr); 
+        detail::character_flags_table_type flag = pair_char_info.second;
+
+        if( (itr == d_str.begin()) ? IS_LOWER(flag) : IS_UPPER(flag) )
+          the_chr = convert_char(pair_char_info);
+        d_buffer += detail::from_char_utf8(the_chr, d_buffer);
```

Comment:
The `probe` and `execute` functors differ by only a couple of lines of code. What about combining them into a single functor? The `d_buffer` variable is only non-null during execute so here is what capitalize could look like:
```
    __device__ int32_t operator()(size_type idx) {
      if( get_column().is_null(idx) )
        return 0; // null string

      string_view d_str = get_column().template element<string_view>(idx);
      char* d_buffer = get_output_ptr(idx);
      int32_t bytes = 0;

      for( auto itr = d_str.begin(); itr != d_str.end(); ++itr ) {
        auto the_chr = *itr;

        auto pair_char_info = get_char_info(the_chr); 
        detail::character_flags_table_type flag = pair_char_info.second;

        if( (itr == d_str.begin()) ? IS_LOWER(flag) : IS_UPPER(flag) )
          the_chr = convert_char(pair_char_info);

        if( d_buffer ) // only non-null during execute
          d_buffer += detail::from_char_utf8(the_chr, d_buffer);
        bytes += detail::bytes_in_char_utf8(the_chr);
      }
      return bytes;
    }
  };
```
This does both the probe and execute in the same function. The `bytes` are computed in both passes but is just discarded for the execute step. Or you could add an `else` clause for the `bytes += ...` line too.

This would leave one base class plus one functor for capitalize and one functor for title.
The new combined functor could have a ctor like the base class taking the offsets and chars pointers as optional parameters:
```
     probe_execute_capitalize(column_device_view const d_column,
                              int32_t const* d_offsets = nullptr,
                              char* d_chars = nullptr )
             : probe_execute_base(d_column, d_offsets, d_chars) {}
```

Reply:
I don't really see a benefit in doing so. The goal is not to decrease the number of functors at the cost of inflating the functors' operators. This would make the code hard to read and maintain. 

As stated before, merging two pieces of functionality into one breaks cohesion. The functors as they are right now have very clear responsibilities. Cocktailing them together has no clear benefits, just making things confusing for later developers maintaining or extending this functionality.

Reply:
Ok. Combining these is not a requirement.
Mostly I'm concerned someone will try to change the execute function for some feature without realizing they likely need to change the size calculation too.

Reply:
I think I fixed the sync requirement between probe and execute. Essentially, I isolated the logic, `generate_chr()` to be called inside the `operator()`'s loop so that each probe/execute pair calls the same logic for capitalize and for title, respectively. Since a picture is worth a thousand words, see the diagram attached below.

![string_capit_title](https://user-images.githubusercontent.com/37386037/75308739-ccfdd980-5814-11ea-8617-4ee8b5ac7f9f.jpg)

Reply:
This seems overly complex considering what these two simple functions do. 
I now think the original separation would be easier to follow and maintain.